Code/Framework/AzNetworking/AzNetworking/TcpTransport/TcpConnection.cpp

```diff
@@ -328,8 +328,16 @@ namespace AzNetworking
 
         if (dstData == nullptr)
         {
-            AZLOG_ERROR("Send ringbuffer full, dropped packet");
-            return false;
+            // Ring buffer is full, try sending existing data then reserving the ring buffer memory again.
+            UpdateSend();
+            dstData = reinterpret_cast<uint8_t*>(m_sendRingbuffer.ReserveBlockForWrite(headerSize + payloadSize));
+
+            // Still failed to get ring buffer memory, this packet can't be sent.
+            if (dstData == nullptr)
+            {
+                AZLOG_ERROR("Send ringbuffer full, dropped packet");
+                return false;
+            }
```

Comment:
As you pointed out in your description, it'd be nice to have some kind of exponential backoff retry logic here, but this is a good improvement.

Reply:
I like the idea of force sending to clear the buffer but I don't think this is the place to do it. We're seeing this happen due to an exceptional case of the Editor flooding the connection with level data for a massive level. I think the thing to do here is check failure on `SendReliablePacket` (which the return false here will prop up to) and then call `AZ::Interface<AzNetworking::INetworking>::Get()->ForceUpdate();` to flush.

Reply:
> So the issue here is less the Editor getting bogged down but more that a TcpConnection in AzNetworking has a max size to its send buffer and MultiplayerEditor doesn't particularly know what to do when the level exceeds that size. The instinct to flush outstanding data is correct. Presently, when we're sending a massive amount of data the caller is responsible for flushing as needed.
> 
> I also agree with the changes around initialization timing. I'd definitely like to see those submitted.

It's hard to separate the two concerns with the way the system works right now though. Both the send and the receive buffers have maximum sizes, and a part of why the send buffer is getting filled up is precisely _because_ the receiving side wasn't processing fast enough due to getting bogged down in Atom initialization. Flushing doesn't solve the problem by itself, because all the sends would still fail due to the receive buffer being full.

Reply:
I agree that the calling code needs to check for errors and handle them, but I disagree that it's _fully_ the responsibility of the caller. This particular problem is caused by the choice of using a ringbuffer implementation inside of TcpConnection instead of infinitely-growing buffers, shared-ptr management, or other strategies, so it seems user-unfriendly to push that knowledge outside of the API to force every caller to deal with that choice.

That said, I can accept that the extra call to UpdateSend() here isn't particularly helpful, since in the case that the send buffer is full because the receiver isn't listening, calling it just one more time instead of inside of some sort of a loop likely isn't going to help. I'll remove it from here, but I think the overall error-handling strategy should be improved in a way that isn't burdensome on every caller.

Reply:
Yeah I agree, there needs to be a bit of a rethink here systemically. Definitely we should get a longer term issue cut to sig-network to rethink both the memory scheme and the error reporting mechanisms here. 

> > So the issue here is less the Editor getting bogged down but more that a TcpConnection in AzNetworking has a max size to its send buffer and MultiplayerEditor doesn't particularly know what to do when the level exceeds that size. The instinct to flush outstanding data is correct. Presently, when we're sending a massive amount of data the caller is responsible for flushing as needed.
> > I also agree with the changes around initialization timing. I'd definitely like to see those submitted.
> 
> It's hard to separate the two concerns with the way the system works right now though. Both the send and the receive buffers have maximum sizes, and a part of why the send buffer is getting filled up is precisely _because_ the receiving side wasn't processing fast enough due to getting bogged down in Atom initialization. Flushing doesn't solve the problem by itself, because all the sends would still fail due to the receive buffer being full.

The Receive side is interesting. I did some testing and it seems to never fill with your changes. It also seems to be the case that UpdateRecv reads a maximum of the maximum packet size off the buffer before proceeding to deserialize and flush all data on the receive buffer. The whole thing need re-evaluation.

Reply:
If you want to see it fill, change "LegacyCommandLineProcessed" back to "CriticalAssetsCompiled" in MultiplayerEditorConnection. This will cause it to once again announce that the server is ready to receieve packets _before_ Atom initializes, so it will start receiving packets just before getting several seconds of blocking loads that prevent it from processing packets.